### PR TITLE
feature: Infer model property types

### PR DESCRIPTION
**Pull request by canvural**
This PR adds ability to infer model property types by scanning the migration files of the application. :fire:

### CODE REVIEW

**Comment by szepeviktor** [comment]
What a feature!

**Comment by canvural** [comment]
I did not test this in a Lumen project, or on a Laravel package. If someone can do so, it'd be great!

File: tests/Features/Properties/ModelPropertyExtension.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+{
+    /** @var User */
+    private $user;
+
+    public function testPropertyReturnType(): int
+    {
+        return $this->user->id;
+    }
+
+    public function testDateReturnType(): ?Carbon
```
Comment (nunomaduro) [inline tests/Features/Properties/ModelPropertyExtension.php]: Why there is optional return type?

File: tests/Features/Properties/ModelPropertyExtension.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
+<?php
```
Comment (nunomaduro) [inline tests/Features/Properties/ModelPropertyExtension.php]: Would be cool have more tests here.

File: src/Properties/SchemaAggregator.php [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+    }
+
+    /**
+     * @param string                $tableName
+     * @param string                $argName
+     * @param PhpParser\Node\Stmt[] $stmts
+     *
+     * @throws \Exception
+     */
+    private function processColumnUpdates(string $tableName, string $argName, array $stmts): void
```
Comment (nunomaduro) [inline src/Properties/SchemaAggregator.php]: It's ok to maintain all of this?

File: tests/Features/Properties/ModelPropertyExtension.php (hunk 1 above)
Comment (canvural) [reply tests/Features/Properties/ModelPropertyExtension.php]: Because `created_at` is nullable.

File: src/Properties/SchemaAggregator.php (hunk 3 above)
Comment (canvural) [reply src/Properties/SchemaAggregator.php]: I think yes. I tested the code on many different Laravel projects. It is working well. Also, psalm Laravel plugin uses this. So if there is any issue we can help each other.

File: tests/Features/Properties/ModelPropertyExtension.php (hunk 2 above)
Comment (canvural) [reply tests/Features/Properties/ModelPropertyExtension.php]: What kind of tests? I couldn't think of anything different. There are tests mainly for normal return types, date casts (Carbon) and casting according to `$casts` property of the model.

I didn't want to write a test for each type (integer, text, json etc.) Cause mainly it's the same code.

Also outside of these cases, I tested the whole branch in many different Laravel projects.

**Comment by muglug** [comment]
@canvural well done for adding!

It's a little weird to see `(c) Nuno Maduro` on code I wrote, though...

**Comment by canvural** [comment]
@muglug Yeah, you are actually right. The only reason I added that is we have the condition in the [Travis](https://github.com/nunomaduro/larastan/blob/master/.travis.yml#L34) file.

I don't know what would be the better solution in this case. I'd definitely want to give you credit. @nunomaduro what do you say?

**Comment by muglug** [comment]
I think that’s quite a strange condition to have, especially for a package that relies on a lot of other people’s open-source contributions.

**Comment by nunomaduro** [comment]
@canvural Let's remove those headers from every file.

**Comment by nunomaduro** [comment]
@muglug It's not that strange, symfony uses this convention a lot. Anyway, I don't like. Let's remove it from every file.

**Comment by canvural** [comment]
@muglug I see you just [recently added](https://github.com/psalm/laravel-psalm-plugin/pull/20) MIT license to psalm Laravel plugin. (I guess just because of this pr) So I guess we can add the MIT license notice to those 3 files in our repo.

**Comment by muglug** [comment]
I think it may have made sense before license templates were so usable, and before everyone started using Composer. Now it’s trivially easy to find out who authored a given line of code, and also who’s most likely to be able to support that code.

**Comment by muglug** [comment]
> So I guess we can add the MIT license notice to those 3 files in our repo.

I'm fine with that – and if you had a note to tag `@muglug` on any changes you make to those files, I'd have the option to make similar changes in my repo.

**Pull request merged**